### PR TITLE
Correctly handle paths containing whitespace character

### LINKED ISSUE #246: `Path.parent` for Native returns null for space-only parent directory
`Path.parent` utilizes `isBlank()` when checking returned value from `dirname`. This is incorrect, as a parent directory can be created as just a space.

```shell
# shell
mkdir " "
```

The following assertion should pass:
```kotlin
assertEquals(" ", Path(" /.").parent?.toString() ?: "NULL")
```

Which is the same behavior as Jvm's `java.io.File`

If `java.io.File` behavior (the underlying on Jvm) is desired to be mimicked for Native, the `dirname` check should be:

```kotlin
val parentName = dirnameImpl(path)

return when {
    parentName.isEmpty() -> null
    parentName == "." -> null
    parentName == path -> null
    else -> parentName
}
```

### PR DESCRIPTION
**Pull request by fzhinkin**
Fixes #246

### CODE REVIEW

**Comment by 05nelsonm** [comment]
I was incorrect in terms of `java.io.File` parity. `null` should **not** be returned when encountering a return value of `.`

My own tests were not removing the trailing slashes when instantiating things.

`./some_path` should always return `.` for a `parent`

File: core/mingw/src/files/FileSystemMingw.kt [hunk 1]
```diff
@@ -38,6 +38,11 @@ internal actual fun basenameImpl(path: String): String {
 
 internal actual fun isAbsoluteImpl(path: String): Boolean {
     if (path.startsWith(SystemPathSeparator)) return true
+    if (path.length > 1 && path[1] == ':') {
+        if (path.length == 1) return false
```
Comment (shanshin) [inline core/mingw/src/files/FileSystemMingw.kt:42]: `path.length > 1` and inside `path.length == 1` ?!

File: core/native/src/files/PathsNative.kt [hunk 2]
```diff
@@ -16,17 +16,19 @@ import platform.posix.*
...
+    rawPath: String,
     @Suppress("UNUSED_PARAMETER") any: Any?
 ) {
+    internal val path = removeTrailingSeparators(rawPath)
+
     public actual val parent: Path?
         get() {
             when {
-                path.isBlank() -> return null
+                path.isEmpty() -> return null
```
Comment (shanshin) [inline core/native/src/files/PathsNative.kt:27]: why aren't checks `!path.contains(` used like in JS?

Comment (fzhinkin) [reply core/native/src/files/PathsNative.kt:27]: On native, that check is performed by corresponding `dirnameImpl` implementations. 
For example, https://github.com/Kotlin/kotlinx-io/blob/fe9eefc72eb2f500ab7f1f0cd5f7d1baa351d815/core/linux/src/files/FileSystemLinux.kt#L17
On all native platforms except MinGW, it checks `/` presence, but on MinGW, it also checks for `\` presence. For that reason, the check is inside `dirnameImpl` and not inside `parent`.

File: core/mingw/src/files/FileSystemMingw.kt (hunk 1 above)
Comment (fzhinkin) [reply core/mingw/src/files/FileSystemMingw.kt:42]: Seems like I need more tests

File: core/js/test/files/SmokeFileTestWindowsJS.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+package kotlinx.io.files
+
+import kotlinx.io.isWindows
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertNull
+
+class SmokeFileTestWindowMinGW  {
```
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/js/test/files/SmokeFileTestWindowsJS.kt:13]: Oh, sorry :(

[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/js/test/files/SmokeFileTestWindowsJS.kt]: Added cases covering this particular bug

File: core/js/src/files/PathsJs.kt [hunk 4]
```diff
@@ -84,6 +72,8 @@ public actual val SystemPathSeparator: Char by lazy {
     sep[0]
 }
 
+private const val WindowsPathSeparator: Char = '\\'
```
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/js/src/files/PathsJs.kt]: Done

File: core/common/src/files/Paths.kt [hunk 5]
```diff
@@ -128,3 +128,22 @@ public fun Path.source(): Source = SystemFileSystem.source(this).buffered()
 )
 @JvmName("sinkDeprecated")
 public fun Path.sink(): Sink = SystemFileSystem.sink(this).buffered()
+
+internal fun removeTrailingSeparators(path: String, sep: Char): String {
```
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/common/src/files/Paths.kt]: Moved everything to a common source set. Now it's both easier to follow when reading the code and easier to test.

File: core/common/src/files/Paths.kt [hunk 6]
```diff
@@ -129,20 +129,46 @@ public fun Path.source(): Source = SystemFileSystem.source(this).buffered()
...
+                2
+            } else {
+                1
+            }
+        } else {
+            1
+        }
+        return removeTrailingSeparatorsWindws(limit, path)
+    }
+    return removeTrailingSeparatorsUnix(path)
```
Comment (shanshin) [inline core/common/src/files/Paths.kt:148]: Is it correct for all native platforms to check on `UnixPathSeparator`?
Shouldn't there be an `SystemPathSeparator` here?

Comment (fzhinkin) [reply core/common/src/files/Paths.kt:148]: Only on Windows both `UnixPathSeparator` and `WindowsPathSeparator` could be used interchangeably (or at least the choice of separator in a particular path won't affect how it is treated by the FS). On all other platforms the only separator is `UnixPathSeparator`.

Comment (shanshin) [reply core/common/src/files/Paths.kt:148]: but we only go into the Windows branch in tests, in other cases we always end up here

**Review by shanshin — APPROVED** [review]

**Pull request merged**